File: frontend/src/api/goals.ts
import { apiClient } from './client';

export interface LearningGoal {
  goal_id: string;
  topic: string;
  target_mastery: number;
  target_date: string;
  current_mastery: number;
  progress_percentage: number;
  days_remaining: number;
  on_track: boolean;
}

export async function getGoals(learnerId: string) {
  const { data } = await apiClient.get<LearningGoal[]>(`/goals/${learnerId}`);
  return data;
}

export async function createGoal(
  learnerId: string,
  body: { topic: string; target_mastery: number; target_date: string },
) {
  const { data } = await apiClient.post(`/goals/${learnerId}`, body);
  return data;
}

export async function deleteGoal(learnerId: string, goalId: string) {
  await apiClient.delete(`/goals/${learnerId}/${goalId}`);
}

File: fastapi_app/routers/goals.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException

from fastapi_app.schemas.platform import GoalCreateRequest
from fastapi_app.security import require_api_key
from fastapi_app.services import goals_service

router = APIRouter(prefix="/goals", tags=["Goals"])


@router.get("/{learner_id}")
def list_goals(learner_id: str, _: None = Depends(require_api_key)):
    return goals_service.list_goals(learner_id)


@router.post("/{learner_id}")
def create_goal(learner_id: str, payload: GoalCreateRequest, _: None = Depends(require_api_key)):
    return goals_service.create_goal(
        learner_id, payload.topic, payload.target_mastery, payload.target_date
    )


@router.delete("/{learner_id}/{goal_id}")
def delete_goal(learner_id: str, goal_id: str, _: None = Depends(require_api_key)):
    ok = goals_service.delete_goal(learner_id, goal_id)
    if not ok:
        raise HTTPException(status_code=404, detail={"detail": "Goal not found", "code": "not_found"})
    return {"ok": True}
